### ISSUE TITLE
Implement VST support for Linux

### ISSUE THREAD

**Issue by cbjeukendrup** [musescore/MuseScore#16799]
**Task description**
On Linux, VST support was delayed due to critical issues with it (mainly the GUI not working because we need to provide the plugin with an event loop). The plan is to implement this for 4.1. Contributions from the community are very welcome.

For some context, see:
#11689
https://github.com/musescore/MuseScore/issues/11592
https://github.com/musescore/MuseScore/issues/11550
https://github.com/musescore/MuseScore/issues/9481

**Comment by cbjeukendrup** [comment]
Yesterday I investigated a bit what the problem here actually is. I found this page from the Steinberg documentation: https://steinbergmedia.github.io/vst3_dev_portal/pages/Technical+Documentation/Provide+A+Runloop+On+Linux/Index.html

> On Linux, there is no global event run loop like on Windows and macOS. For this reason, the plug-in can query for an [Linux::IRunLoop](https://steinbergmedia.github.io/vst3_doc/base/classSteinberg_1_1Linux_1_1IRunLoop.html) which must be provided by the host application in two different ways.

It would of course be great to make use of the runloop provided by Qt. On a first sight, that seems impossible, because for Steinberg's IRunLoop interface we need to implement something with a _file descriptor_ while you usually see nothing like that in Qt's event loop, which is much more high-level.
However, I found this class: https://doc.qt.io/qt-6/qsocketnotifier.html That might be exactly what's needed to implement the first two methods of the IRunLoop interface. 
For the later two methods, I think we can just use a QTimer. 

Some resources that might be helpful:
- Audacity's implementation: https://github.com/audacity/audacity/blob/master/src/effects/VST3/internal/x11/RunLoop.cpp
- A very minimal host example from Steinberg: https://github.com/steinbergmedia/vst3_public_sdk/blob/3f20bcbf36283b1e5c450e088571be6226cd4e64/samples/vst-hosting/editorhost/source/platform/linux/irunloopimpl.h

However, I think both of these implementations are far more complicated than what ours will end up being, because we have the possibility to use stuff from Qt.

**Comment by bmarwell** [comment]
Have you tried talking to someone from the KDE project? They might implement something for you in future versions. Of course, this will not be helpful in the near future, but hopefully in the long run.

**Comment by cbjeukendrup** [comment]
I don't really think we need support from KDE. What needs to happen is simply that we need to implement some interface from the VST SDK, so that it can use that interface to set up event handling and timers. And my hypothesis is that we already have everything that's needed to implement that interface properly, namely via Qt.
Or did you mean they might want to help us implement it _in MuseScore_? That would be very nice of course, but I think we don't need to actively ask them to spend their time on that.

**Comment by bmarwell** [comment]
Nah, first one. I understood some platform APIs were missing and I just wanted to mention that reaching out is probably an option. I just wondered it was not mentioned before, that's all. 😃

**Comment by bmarwell** [comment]
qt and KDE are historically tied together. Sorry if that was not the case anymore.

**Comment by ThePython10110** [comment]
For now, if you don't want to build from source with the VST flag, you can use [@diedeno's builds](https://github.com/diedeno/MuseScore/releases), which does the same thing.

**Comment by cbjeukendrup** [comment]
Finished in https://github.com/musescore/MuseScore/pull/28382. Problems with specific plugins can be reported separately, if they haven't been reported already.

**Comment by ShawkMusic** [comment]
> Finished in [#28382](https://github.com/musescore/MuseScore/pull/28382). Problems with specific plugins can be reported separately, if they haven't been reported already.

YOOOOOOOO!!!
Will this be in the next release of Musescore? If so, any idea when that might be?
Also I'm assuming by default it'll only work with native programs, right? How well does it work with compatibility programs like Yabridge?

**Comment by cbjeukendrup** [comment]
It will be in MuseScore 4.6, expected in September. 
In principle, all VST3 plugins should work. I myself can't comment about particular ones, but perhaps someone else knows. 
And as said, feel free to report issues with particular plugins or in particular situations; we'll see what can be done about those!

**Comment by diedeno** [comment]
It works with *most* native Linux vst3's, as well as with *most* windows vst3 that work with wine(+yabridge or other) to start with.
